### ISSUE TITLE
Rendering to 99.98% and never ending

### ISSUE THREAD

**Issue by SomeBastardTookMyUserName** [buddhi1980/mandelbulber2#444]
Howdy!  Thanks so much for this software, buddhi!  I wrote about this problem on fractalforums.org and was advised to write up a big report here.  I would love for you to tell me I'm an idiot and I've done something stupid, but I fear there may actually be some issue here. 

##### System information (version)
- mandelbulber2 => 2.11-1
- Operating System / Platform => OSX 10.11.5
- Compiler => None, prebuilt dmg

##### Detailed description
I'm having issues trying to render animation from mandelbulber on several machines at once.  My cluster of old-ass Macs is connected via Ethernet and normally runs mb3d via the open source render farm manager Dr. Queue. However, I just ran exactly the same test using mb3d's own network render protocol, so now I'm convinced there's something else amiss.

The attached settings file will render fine as any given single image on a single machine. The file can be saved from that single machine to any other drive on the network.  So read/write permissions and network protocol seems OK.

Any animation render results in multiple instances of mb3d opening, seeming to render, showing a progress rate of 99.98%, and then hanging.  Forever.  The machines were left alone for three weeks, and they were still hung up.  Nothing is written to disk.

This has been consistent for TWO separate settings files with two different sets of fractal data, although they were both mandelboxes, if that helps to know.  Both settings files were also attempting stereoscopic equirectangular renders, with around 300 keyframes, asked to render 100 frames per keyframe.  No collisions detected when validating.  

On the farm the render manager would deploy 40 cores at a time and 40 instances would begin, all hang at 99.8% and no data would be written.

Render farm software is issuing this command to the client machines:
/PATH_TO_MANDELBULBER_APP /PATH_TO_INPUT_FILE -K -s [start frame] -e [end frame] -f png16alpha -o /PATH_TO_OUTPUT

This script actually HAS worked fine on 10,000 frames of a single POV fractal zoom on the exact same machines using mb3d v. 2.08.  Regardless of the render farm software, as I mentioned, mb3d's own network render fails exactly the same way.

Variables tried, no discernable effect:
switching out png16alpha for exr, png, jpg
switching out output locations to local directories, desktop, other mounted volumes
switching out fractal settings files

##### Steps to reproduce
Attached Mandelbulber file is not the latest culprit, but certainly a culprit.  

[BoxZoom3.fract.zip](https://github.com/buddhi1980/mandelbulber2/files/1669737/BoxZoom3.fract.zip)

**Comment by zebastian** [comment]
hey @SomeBastardTookMyUserName.
Can you try to decrease the screen resolution to something fastly testable (for example 100x100) and see if the issue is still present?
I could not recreate the issue yet...

**Comment by SomeBastardTookMyUserName** [comment]
Yes, I'll try that today.

**Comment by SomeBastardTookMyUserName** [comment]
OK, here are some interesting results!  Well, they are all fails, but it still may tell you something.

1. I did try the following script on the Render Farm software to see if there were any problems with network, permissions, writing to disks, or a general inability of the render farm system itself, and not Mandelbulber.  It passed, though, so it seems like all those issues are not issues.  It was just this:

`cp /Volumes/Lucifer/Script\ Archive/_template.txt ~/Desktop`

"Lucifer" being a mounted volume.  Every node on the farm copied the file successfully with no issue.

2. I resized the .fract file to 320x240 as per your instructions, and tried that on the farm.  I include at the end the log file from this experiment as well as a screenshot of the terminal output showing the hang at 99.98%.  The new 320 .fract file also.

3. I tried to render a single frame from the Terminal using this command:
`/Applications/mandelbulber2.app/Contents/MacOS/mandelbulber2 /Volumes/Lucifer/Tests/Mandelbulber/New_Box_320.fract -n -K -s 15 -e 16 -f png16alpha -o /Users/Shared/Sinistar/360/`

Where the first line is the path to Mandelbulber, the second argument is the .fract file, and the last argument is the path to a hardlink for a networked volume.  Bear in mind that the same designations and hardlinks work well for the cp example above.  Frame 15 is arbitrary, and there are no collisions in the file.  This also failed, hanging at 99.98%.

Any help is appreciated!

[Archive.zip](https://github.com/buddhi1980/mandelbulber2/files/1672630/Archive.zip)

**Comment by SomeBastardTookMyUserName** [comment]
As of yesterday, tried to render via command line on OSX 10.10 (Yosemite) machine with same settings.  Stalled at 99.98%.  Only from command line, it renders the same frame via the GUI with no problems.

Same test on OSX 10.9.5, Mandelbulber says it rendered everything, but does not write any file to disk.  Tried both leaving name blank and pointing the output to a directory and declaring the filename as "test.png."

I have a feeling it's just me doing something unbelievably stupid.

**Comment by SomeBastardTookMyUserName** [comment]
I'm not.  Going back to version 2.08-3 and attempting the exact same set of procedures produces nothing but rousing successes, even when I fake out a 2.11 settings file and change it to say "2.06" in the header.  Works fine.

I know y'all are working on more important things for version 2.12, and I'm glad to see that.  But command line rendering on OSX for v. 2.11 does seem to be at issue here, and there's no OSX build for 2.12 yet.  I'll try to build it on my end, because if it still doesn't work in 2.12 then you might have a real issue maybe you should be looking into while you're doing 12.

**Comment by mclarekin** [comment]
The person who has been in charge of making the OSX builds for the last few
versions is no longer able to help us.

So in summary, I am reading that some where between 2.08-3  and  2.11,
there has been some code change that has broken the command line rendering.
Do you know if it worked in 2.09 and 2.10?

**Comment by SomeBastardTookMyUserName** [comment]
I'll admit I went to the last one in which I KNEW it worked.  I can do those tests easily, and I'll report back.  Thanks as usual for the assistance, mclarekin!

**Comment by SomeBastardTookMyUserName** [comment]
P.S. I'm not a programmer by any means, but I CAN build from source if I'm given instructions.  If I can be in any way helpful towards getting macos versions out, just let me know what I can do.  Pretty much a greenhorn when it comes to this, though.  No way I could trace the problem on my own or I would have done so!

**Comment by SomeBastardTookMyUserName** [comment]
Tests done:
Current source will not build on either 10.10 Yosemite and 10.9.5 Mavericks. in each of these cases the dependencies seem fine but when I go to qmake the application it errors out asking for QtWidgets.  Google that and you'll get a note suggesting I need to have QT 5.0 or later installed, but that won't install on 10.10 or 10.9.5.  - I'm limited to v 4.8.

So I try 10.12.6 Sierra.  The build bombs before I could get dependencies working - libpng fails with the error ZLIB_VERNUM != PNG_ZLIB_VERNUM "-I (include path).  Installing the latest zlib does not seem to fix that; what do I know about this stuff.

Command-Line render on version 2.09 of Mandelbulber stalls again at 99.98% and goes no further.  So yes, it seems to be something that happened right after version 2.08.

I guess I'll slush back to version 2.08 and do my render without stereo!  Kind of a pity, that.  Unless one of y'all has a good solution to this.  I'll slam my head against any wall for a while to get it going.

**Comment by mancoast** [comment]
Are you installing deps with “brew” ?

**Comment by SomeBastardTookMyUserName** [comment]
No, via the instructions included in the distribution.  One via ftp, one via http, one via curl.  Should I?  Can certainly try that today.

**Comment by mancoast** [comment]
Check this yaml file for build steps: https://github.com/buddhi1980/mandelbulber2/blob/master/circle.yml

I would highly recommend using brew to install deps.

**Comment by zebastian** [comment]
@mancoast @SomeBastardTookMyUserName yes, brew seems to be the package manager of choice for build deppendencies on macos.
i will also some more tests with animation-problem, when im back (around end of february)

**Comment by SomeBastardTookMyUserName** [comment]
Yes, I adore brew with all my heart and soul!  So easy.  I was just following instructions marked "OSX"  included in the source.  Thanks so much for the .yml file - I am readying a clean machine for the build according to those specs now.  Will report how it's going, and am newly hopeful.

In the meantime it may please y'all to note that the 2.08 version is newly deployed on my li'l trashnet render farm and the file is busily humming along with 40 cores rendering 8K equirectangular files right now.  I do love this software.

**Comment by GordonHarley** [comment]
I can confirm seeing something similar when setting optimization for high quality (without my modifications in place ;0), where the estimated raymarching step level continually approaches a zero limit but never completes.  Repro is setting opencl engine to full, and turning on High optmization with menger sponge or dodecahedron.   I'm on a vega64 and the drivers aren't mature yet, so this could be an issue with the drivers, but I've tried on two versions.   

If I let it run, it usually causes a driver restart.  

That apparently happens on AMD & NVidia cards and cause seems to be the drivers, but I'm more concerned about step level constantly moving towards zero.

**Comment by buddhi1980** [comment]
@GordonHarley , problem mentioned by you is different than main topic. Which version do you use? In actual development version it is already corrected.

**Comment by mclarekin** [comment]
My two cents,a bit off the topic:

a)  does macos have GPU time out issues that require restart,  like in
Windows where users have to edit the registry to use openCL in Mandelbulber
and GLSL Fragmentarium??.  High Quality (= small DE steps)  can cause
driver restart on windows if the registry is not edited

b)  i do not use the quality settings optimization options, it is faster to
control the quality manually.
And  when I did try optimization a few versions ago, it occasionally got
caught in an increasing loop, so the DE step multiplier increased to a
maximum big number (weird). I am assuming this has probably been fixed by
now.
I only rarely go below step multiplier 0.1 and probably never below 0.001.
Optimization of quality control is a good tool when the fractal maths and
the Distance Estimate are nicely in sink. With weird setups it is
sometimes  best to do it manually

**Comment by SomeBastardTookMyUserName** [comment]
Yeah, I considered the ray-marching issue a LOOONG time ago, because that would be an obvious one, even to me.  But the key issue I'm reporting is this:

Command-Line rendering stalls at 99.98% and never finishes.

The exact same frame will render JUST FINE in GUI mode.  Exact same, same settings file, same everything.

Also: build report for mancoast and anyone else who cares.   Strict adherence to the .tml file ended up in no build.  The errors went like this.
```
Project MESSAGE: Use QtGamepad module
Project MESSAGE: Use tiff library for TIFF files
Project MESSAGE: Use sndfile library for WAV files
Empty filename passed to function
Empty filename passed to function
```
Repeat for about 100 times, then, when make is executed I got the usual 500 lines of log with lots of errors at the end, all of which said:
`...means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.`
All well and good, I suppose, so I opened up XCode, had it fix those errors, and had the XCode GUI build the binary!  Sweet success!

Ooops. Well, maybe not quite.  Renders from the GUI come out as solid grey boxes with nothing in them.  OK, maybe my settings file is not so good.  Here's the damning issue again.  Command-line render stalled at 99.98%, never finished:
```
tmaloney$ /Users/tmaloney/mandelbulber2/mandelbulber2/qmake/Debug/mandelbulber2.app/Contents/MacOS/mandelbulber2 /Users/tmaloney/Desktop/New_Box.fract -n -K -s 15 -e 16 -f png16alpha -o /Users/tmaloney/Desktop 
Mandelbulber 2.12
Log file name: /Users/tmaloney/.mandelbulber_log.txt
Detected 4 CPUs
Default data hidden directory: /Users/tmaloney/.mandelbulber/
Default data public directory: /Users/tmaloney/mandelbulber/
Initialization: Setting up image buffers [                                    ]
Initialization: Loading textures [                                            ]
Frame 15 of 30000 (key 0) Done 0.00%, elapsed: 0.3s, estimated to end: n/a [  ]
Rendering image: Done 99.98%, elapsed: 16m 46.8s, estimated to end: 0.1s [### ]
^C
```
Full giant-ass ugly logs available on request, but it sounds like I've driven this one into the ground until better help comes along.  As far as I can tell, command-line rendering is broken after v 2.08 and is still not really fixed.  Thank you all for listening!

**Comment by bermarte** [comment]
I can have a look to it these days.

**Comment by zebastian** [comment]
@SomeBastardTookMyUserName @bermarte we now have compiled dmg files produced by circle ci, here is a link to the most recent stable version: https://289-33041722-gh.circle-artifacts.com/0/Users/distiller/mandelbulber2/build/artifacts/mandelbulber2-macOS-7d50dc86836dbeb61595402b792ae8398b201364.dmg
Could you try running these and @SomeBastardTookMyUserName can you check if the progress issue is fixed in this version?

**Comment by SomeBastardTookMyUserName** [comment]
Hey, it's me!  Some Bastard Took My User Name!  Back with a proper identity.

I tuned out to finish my render, a 30,000 frame 360 fractal move.  Just finished last week, so I was preoccupied.  I am thrilled to see this new build, because I could not make that render stereo and that was my first choice.

Thanks for posting, sorry I'm getting to this a month late.  I apologize if I seem ungrateful.  WIll test as soon as I can!

**Comment by SomeBastardTookMyUserName** [comment]
Although, oops, I just saw I can't boot this up on OSX 10.9.  Yeah, yeah, I know, that's old.  But for stability purposes that's what I'm still on.  The farm is on 10.11, so I hope this is not compiled ONLY for the latest, or I won't really be able to test until I set a new machine up.  And then it won't be available for me, because upgrading the farm would be a dumb idea at this point.

But I'll certainly get some data.
